_posts/2023-11-16-quarkus-meets-langchain4j.adoc

```diff
@@ -0,0 +1,265 @@
+        return find("id", id).firstResult().name;
+    }
+
+}
+----
+
+In this example, we are using the Panache repository pattern to access the database. We have a specific method annotated with `@Tool` to retrieve the customer name. When the LLM needs to get the customer name, it instructs Quarkus to call this method and receives the result.
+
+Obviously, it's not a good idea to expose every operation to the LLM. So, in addition to `@Tool`, you need to list the set of tools you allow the LLM to invoke in the `@RegisterAiService` annotation:
+
+[source, java]
+----
+@RegisterAiService(
+    tools = { TransactionRepository.class, CustomerRepository.class },
+    chatMemoryProviderSupplier = RegisterAiService.BeanChatMemoryProviderSupplier.class
+)
+public interface FraudDetectionAi {
+   // ...
+}
+----
+
+
+The `chatMemoryProviderSupplier` configuration may raise questions. When using tools, a sequence of messages unfolds behind the scenes. It becomes necessary to configure the AI service's memory to adeptly track these interactions. The `chatMemoryProviderSupplier` offers a solution by enabling you to define how the memory is handled. The `chatMemoryProviderSupplier` lets you define how the memory is managed. The `BeanChatMemoryProviderSupplier` looks for a `ChatMemoryProvider` bean, like the following:
+
+[source, java]
+----
+@RequestScoped
+public class ChatMemoryBean implements ChatMemoryProvider {
+
+    private final Map<Object, ChatMemory> memories = new ConcurrentHashMap<>();
+
+    @Override
+    public ChatMemory get(Object memoryId) {
+        return memories.computeIfAbsent(memoryId, id -> MessageWindowChatMemory.builder()
+                .maxMessages(20)
+                .id(memoryId)
+                .build());
+    }
+
+    @PreDestroy
+    public void close() {
+        memories.clear();
+    }
+}
+----
+
+
+IMPORTANT: Note that not all LLMs support tools.
+
+=== Document stores
+
+Document stores are a way to extend the LLM knowledge with your own data. This approach - called  _RAG_ - requires two processes:
+
+First, you need to ingest documents into a document store. The documents are not stored as-is, but an embedding is computed. This embedding is a vector representation of the document.
+
+Then, in the Quarkus application, you need to declare the document store and the embedding to use. The extension will load the document store and the embedding and will use them to extend the LLM knowledge. Thus, before calling the LLM, it retrieves the relevant documents from the store (that's where the vector representation is useful) and attaches them to the LLM context (which essentially means adding the retrieved information from the document to the user message).
+
+The Quarkus Langchain4j extension provides facilities for both steps.
+
+The following code shows how to ingest a document into a Redis document store:
+
+[source, java]
+----
+@ApplicationScoped
+public class IngestorExample {
```

Comment:
Could this  be linked to the previous example somehow?

Reply:
WDYM?

Reply:
Well, we have a very good example above that allows to understand exactly why we have these concepts and this code here is disconnected from that so it's harder to imagine how it could be useful.

But I'm not sure if we could have a usage for this concept in the previous triaging example.